File: src/Main/java/com/nishant/algorithms/GraphTheory/ElementaryGraphAlgorithms/BreadthFirstSearch/BreadthFirstSearch.java
package com.nishant.algorithms.graphtheory.elementarygraphalgorithms.breadthfirstsearch;

import com.nishant.algorithms.graphtheory.elementarygraphalgorithms.Utility;
import com.nishant.algorithms.graphtheory.implementations.Color;
import com.nishant.algorithms.graphtheory.implementations.Pair;
import com.nishant.algorithms.graphtheory.implementations.Vertex;
import com.nishant.algorithms.graphtheory.implementations.adjacencylist.DirectedGraphAdjacencyList;

import java.util.LinkedList;
import java.util.Queue;

public class BreadthFirstSearch {

  public static <E> void search(DirectedGraphAdjacencyList<E> graph, E source) {
    Utility.initializeTraversal(graph);
    Queue<Pair<Vertex<E>, Integer>> queue = new LinkedList<>();
    queue.offer(new Pair<>(graph.get(source), 0));
    int depth = 0;

    while (!queue.isEmpty()) {
      Pair<Vertex<E>, Integer> next = queue.poll();
      Vertex<E> nextVertex = next.first;
      int nextDepth = next.second;
      nextVertex.color = Color.grey;

      if (depth != nextDepth) {
        System.out.println();
        depth++;
      }

      nextVertex.dist = depth;

      System.out.print(nextVertex + " ");

      for (Vertex<E> neighbor : graph.adjacencyList.get(nextVertex)) {
        if (neighbor.color == Color.white) {
          queue.offer(new Pair<>(neighbor, depth + 1));
        }
      }

      nextVertex.color = Color.black;
    }

    System.out.println();
  }
}

File: src/main/cpp/graphtheory/traversals/BreadthFirstSearch.cpp
#include "BreadthFirstSearch.h"
#include "../GraphMaker.h"

std::vector<int> breadthFirstSearch(std::vector<std::list<int>> graph, int start) {
  enum State {WHITE, GREY, BLACK};
  std::vector<State> states(graph.size(), State::WHITE);

  std::queue<int> q;
  q.push(start);

  int next;
  std::vector<int> res;  

  while (!q.empty()) {
    next = q.front();
    q.pop();
    states[next] = BLACK;
    res.push_back(next);

    for (int neighbor : graph[next]) {
      if (states[neighbor] == WHITE) {
	states[neighbor] = GREY;
	q.push(neighbor);
      }
    }
  }

  return res;
}

int main() {
  std::vector<std::list<int>> graph = makeGraphAdjList(5, {{0, 1}, {1, 2}, {2, 3}, {3, 4}, {0, 4}, {1, 4}});

  std::vector<int> searched = breadthFirstSearch(graph, 0);
  for (auto it = searched.begin(); it != searched.end(); ++it) {
    std::cout << *it << " ";
  }

  std::cout << std::endl;
}
